### DISCUSSION TITLE
Framer Motion + NextJs crash module styles when animate page leaving

### DISCUSSION THREAD

**Discussion by Mistwell** [vercel/next.js#18724] [Help]
Next Js >= 9.5.5 + Framer Motion
When change page route, old styles dissapear while playing exit animation. Not classes or generated class names, but just styles in class.

Next v9.4.5 works perfect

![Без названия](https://user-images.githubusercontent.com/45655927/97965972-9b420700-1dd4-11eb-958e-0b53cf20574f.png)
![Без названия (1)](https://user-images.githubusercontent.com/45655927/97965977-9d0bca80-1dd4-11eb-8959-46ce8970039f.png)

**Comment by qroux** [comment]
I think i'm facing a similar issue. ( **next: 9.5.5 | framer-motion: 1.11.1 | sass: 1.27.0** ). Framer exit animation was working great until I have decided to refactor my css using sass and the built-in module (Component.module.scss).

Now, when I navigate between pages, it seems like:
- **the leaving page is losing its styling instantly**
- **while the leaving-animation from Framer is running correctly** (with the default global.scss styling instead of its own specific styling)

_

> Edit: this only happens in production mode (after next build) and not in dev mode.

_

At this point, I don't know if:
- the issue is related to **the build process**
- I forgot some configuration to tell next or framer **to not remove the current styling before the end of the leaving animation**
- this is the result of an incompatibilty between Framer-motion and Next sass module 


My code:

    <motion.div
      initial={{ opacity: 0, y: '30px' }}
      animate={{ opacity: 1, y: '0px' }}
      exit={{ opacity: 0, y: '30px' }}
    >
      <div className={styles.index__layout}>
        <div className={styles.index__container}>
          <h1 className={styles.index__heading}>Orders History</h1>
          <div className={styles.index__identification}>
            <h4 className={styles.index__identification__heading}>Account: <strong>{currentUser.email}</strong></h4>
          </div>
          <div className={styles.index__orders}>
            {orderList}
          </div>          
        </div>
      </div>

    </motion.div>

**Comment by qroux** [comment]
The issue is related to the building process of the Component.module.css.

**same issue** https://github.com/vercel/next.js/issues/16723

**Solutions:**
- Downgrade your next version to 9.5.0 until they have fixed it.
- Use inline styling
- Use Css-in-Js like "styled-components"

**Comment by OlivierFortier** [comment]
I have been having the exact same issue... Sadly I have upgraded to next 10 to use the new Image component... It's going to be tough to go back with no automatic image optimisation

**Comment by rlski** [comment]
I've also been facing the same issue with next > 9.5.0, sass-modules et framer. I'd love to be able to use this stack with Next 10, it's really nice to use.

**Comment by le-martre** [comment]
Same issue for me ! :( And also bummed about not being able to use the new Image Component

**Comment by slhnx** [comment]
Same issue for me in NextJS. The page looses it's styling when page transition happens. (Only in production). Is there a solution for this problem in the new nextjs canary ?

**Reply by yannherren** [reply]
I've got the same issue in my project. It's possibly because the CSS modules get unloaded as soon as the route changes. I've found a possible workaround on this site, but I'm not if it's going to work, haven't tried it yet. It's about loading the CSS Modules globally. Maybe it can help.
https://bleepcoder.com/next-js/730844449/css-is-reset-on-route-change-in-production-build

**Comment by yannherren** [comment]
I'm experiencing the same issue. It makes framer unusable with NextJS. Would also be grateful to have a fix for this.

**Comment by kharrat** [comment]
I found a post on Stackoverflow https://stackoverflow.com/questions/65574151/css-module-being-removed-on-path-change-before-framer-motion-exit-animation-comp and edited a bit, this work for me:

```jsx
// _app.js

import { useEffect } from "react";
import { AnimatePresence } from "framer-motion";
import Router from "next/router";

function MyApp(props) {
  const { Component, pageProps, router } = props;

  let copies = [];

  useEffect(() => {
    Router.events.on("beforeHistoryChange", onLoad);
    return () => {
      Router.events.off("beforeHistoryChange", onLoad);
    };
  }, []);

  const onLoad = () => {
    // Create a clone of every <style> and <link> that currently affects the page. It doesn't matter
    // if Next.js is going to remove them or not since we are going to remove the copies ourselves
    // later on when the transition finishes.
    const nodes = document.querySelectorAll(
      "link[rel=stylesheet], style:not([media=x])"
    );
    copies = [...nodes].map((el) => el.cloneNode(true));

    for (let copy of copies) {
      // Remove Next.js' data attributes so the copies are not removed from the DOM in the route
      // change process.
      copy.removeAttribute("data-n-p");
      copy.removeAttribute("data-n-href");

      // Add duplicated nodes to the DOM.
      document.head.appendChild(copy);
    }
  };

  const onExit = () => {
    for (let copy of copies) {
      // Remove previous page's styles after the transition has finalized.
      document.head.removeChild(copy);
    }
  };

  return (
    <AnimatePresence exitBeforeEnter={true} onExitComplete={onExit}>
      <Component {...pageProps} key={router.route} />
    </AnimatePresence>
  );
}

export default MyApp;
```

**Reply by SalmanK81099** [reply]
Hi, this code has a little problem, it does not remove the old CSS, since I am working on quite a large project with a lot of CSS, once I do a couple of page transitions the site becomes unresponsive and crashes, any fix found yet?

**Reply by MatheusLasserre** [reply]
Same problem here

**Reply by abukhalidrifat** [reply]
any solution for this problem. the above code slows down my website.

**Comment by jjaburke91** [comment]
Also occurring for me, regardless if I'm using Framer or React Page Transition to track unmounting components. Styles on exiting component get instantly removed, despite those components having correct ClassName applied to them as by the css modules. Only occurs on build, not in dev. 

Any ideas or solutions on resolving?

Using Next 12.0.8. React v17.0.2 .

**Comment by jjaburke91** [comment]
Found a relevant open issue here: https://github.com/vercel/next.js/issues/17464

**Comment by gdlazcano** [comment]
Anything yet? @kharrat solutions works but I believe it to be inefficient in big projects. The ones mentioned in https://github.com/vercel/next.js/issues/17464 don't fix the issue either

**Comment by harshpandey002** [comment]
Hi, 
Anyone found any fix for this ?
This issue has been there for soo long.

**Reply by julian624** [reply]
i need an asnwer too, need to fix it

**Comment by stentibbing** [comment]
I don't see this as issue with Next.js which is working as expected. Rather I think it's a missing feature of Framer which does ship AnimatePresence but that is rather targeted for React not Next. Since I have a quite large project by now and I do not wish to move to global css, I'll try something I came up with - have a custom Link for next.js which does the actual route change only after exit animation has finished. Feels like the cleanest solution for now.

**Reply by julian624** [reply]
Can you explain it to me please? i need to fix it

**Reply by stentibbing** [reply]
Just did a quick implementation and works like a charm.

In short I created two components, a custom Link to replace next/link and a wrapper which replaces AnimatePresence. The Link creates a custom event which fires on click. The wrapper listens to that event and on event it first executes the exit animation, then executes route change with router.push and finally listens to routeChangeComplete on which it executes the enter animation.

This ensures that the component level css is not removed before exit animation is complete and that enter animation is fired only after entering css is injected.

**Reply by julian624** [reply]
Thank you very much, anyway it's a little difficult for me to understand it, maybe if you don't mind sending an image of how it looks like.

**Reply by stentibbing** [reply]
So first you need a wrapper which does the animating and route pushing. Here's one I threw together:
https://pastebin.com/skMQvFEL

If you used AnimatePresence in your _app.js file to wrapp the Component then you need to import the custom wrapper and replace AnimatePresence with PageTransition.

`
<AnimatePresence mode="wait">
  <Component {...props} />
</AnimatePresence>
` 

Secon part is the Link component. This component replaces the Link component from next.js. Basically it does not do the routing, it just fires an event, asking the wrapper to do routing. You have to use this Link in you're components for navigating between pages. Mine has a bit more functionality: https://pastebin.com/SVJC2Wmg

Hope that helps.

**Reply by stentibbing** [reply]
A small bug I found, in the link you need to pass isExternalLink to the useEffect as an argument. 

And it is worth noting that since we do not start with routing before exit animation, the whole route change will take longer by the duration of the exit animation.

**Reply by NirajD10** [reply]
I may might get error router(importing next/navigation) using next 13.5.6 app router
TypeError: Cannot read properties of undefined (reading 'on') 

lib\customAnimatePresence.js (49:18) @ on

```
47 |     /* Event listeners */
48 |     document.addEventListener("onLinkClicked", handleLinkClick);
> 49 |   router.events.on("routeChangeComplete", handleRouteChangeComplete);
     |                  ^
50 |  
51 |     return () => {
52 |       document.removeEventListener("onLinkClicked", handleLinkClick);
```


so its there any solution for next js 13?

**Comment by KingBael09** [comment]
Hey any updates regarding this....its been soo long?

Initial load works fine (as the css module is available).
However not the same for exit animation (css module removed before animation)

Using Next 13.1.2, React 18.2.0

**Reply by stentibbing** [reply]
Last time I checked (about a month ago), Framers exit animations did not work at all with Next 13, not with App Router atleast.

**Comment by gkpo** [comment]
my god I just lost half a day trying everything. downgrading next, all kinds of workarounds. [kharrat](https://github.com/kharrat) solution worked for me. thanks
